File: leetcode/kotlin/15.3Sum.kt
package com.programming

import java.util.LinkedList

import java.util.Arrays

internal class Test{
   companion object {
       fun threeSum(nums: IntArray): List<List<Int>>? {
           val result: MutableList<List<Int>> = LinkedList()
           Arrays.sort(nums)
           for (i in nums.indices) {
               if (i == 0 || nums[i] != nums[i - 1]) {
                   var start = i + 1
                   var end = nums.size - 1
                   while (start < end) {
                       val sum = nums[i] + nums[start] + nums[end]
                       if (sum == 0) {
                           val tmp: MutableList<Int> = LinkedList()
                           tmp.add(nums[i])
                           tmp.add(nums[start])
                           tmp.add(nums[end])
                           result.add(tmp)
                           val startVal = nums[start]
                           val endVal = nums[end]
                           while (start < end && startVal == nums[start]) {
                               start++
                           }
                           while (end > start && endVal == nums[end]) {
                               end--
                           }
                       } else if (sum < 0) {
                           start++
                       } else {
                           end--
                       }
                   }
               }
           }
           return result
       }
   }
}

fun main(){
    print(Test.threeSum(intArrayOf(-1, 0, 1, 2, -1, -4)))
}

File: leetcode/c++/15.3Sum.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Solution{
  
  public:
  vector<vector<int>> threeSum(vector<int> nums){
     vector<vector<int>> result;
     sort(nums.begin(),nums.end());
     for(int i=0;i<nums.size();i++){
       if(i==0||nums[i]!=nums[i-1]){
         int start = i+1;
         int end = nums.size() - 1;
         while(start<end){
           int s = nums[i] + nums[start] + nums[end];
           if(s==0){
             vector<int> temp;
             temp.insert(temp.end(),{nums[i],nums[start],nums[end]});
             result.push_back(temp);
             int startVal = nums[start];
             int endVal = nums[end];
             while(start < end && startVal == nums[start]){
                   start++;   
             }
             while(end > start && endVal == nums[end]){
                  end--;
             }
           }else if(s<0){start++;}
       else{end--;}
         }
       }
     }
     return result;
  }

};

int main(){
  Solution s;
  vector<int> l{-1, 0, 1, 2, -1, -4};
  for(auto v:s.threeSum(l)){
    cout << "[ ";
    for(auto i:v){
      cout << i <<" , " ;
    }
    cout << " ]" << endl;
  }
  return 0;
}